### ISSUE TITLE
[Help wanted] submit fastfetch to debian unstable

### ISSUE THREAD

**Issue by CarterLi** [fastfetch-cli/fastfetch#533]
Debian and Ubuntu are two of most used Linux distros but we haven't got the official packages there.

**Comment by xzl01** [comment]
我们在deepin的打包的时候做了一些工作，包括将vendor的依赖yyjson单独打包出来，现在正在尝试向debian提交这些包

**Comment by CarterLi** [comment]
@xzl01 @UTsweetyfish Any progress?

**Comment by CarterLi** [comment]
Ref: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1053995

**Comment by xzl01** [comment]
可以参考： https://mentors.debian.net/intro-maintainers/

**Comment by xzl01** [comment]
帮忙联系了一位DD 看是否能够推进下

**Comment by xzl01** [comment]
你可以需要再发一个RFS：https://mentors.debian.net/accounts/register/

**Comment by CarterLi** [comment]
https://mentors.debian.net/package/fastfetch/

Then what?

**Comment by xzl01** [comment]
你需要关注下
![image](https://github.com/fastfetch-cli/fastfetch/assets/31403725/e828b53c-469f-479d-a42f-a5dd97f61d79)
这是对包质量的检查，可能需要对此进行更改

**Comment by xzl01** [comment]
你可以使用  https://manpages.debian.org/testing/lintian/lintian.1.en.html  来检查你的包质量

**Comment by CarterLi** [comment]
https://lintian.debian.org/tags/very-long-line-length-in-source-file.html

这个链接你能打开吗？

**Comment by CarterLi** [comment]
> 你可以使用 https://manpages.debian.org/testing/lintian/lintian.1.en.html 来检查你的包质量

我在用ubuntu。Ubuntu的lintian不报这些错🤣

**Comment by xzl01** [comment]
你可以使用distrobox 运行一个debian容器来跑lintian。  链接我也打不开（可能是debian的问题？）

**Comment by xzl01** [comment]
似乎目前的debian目录依然不符合标准，如果你愿意的话 我给你那个分支提一个pr进行修改

**Comment by CarterLi** [comment]
可以

**Comment by CrazyWolf13** [comment]
@xzl01 and @CarterLi 
Thank you for the amazing work done.

I see there is a [Help wanted] tag, it seems like there is help wanted, but a lot of this issue is on chinese, sadly not everyone speaks chinese, wouldn't english be a bit a better way to communicate?
That way much more people could understand the current situation and possibly help.

**Comment by CarterLi** [comment]
> I see there is a [Help wanted] tag, it seems like there is help wanted, but a lot of this issue is on chinese, sadly not everyone speaks chinese, wouldn't english be a bit a better way to communicate?

I'm aware of this issue and really sorry for that

I had submit ( an old and slightly modified ) fastfetch to debian mentors and got no response. Progress stuck there.

https://mentors.debian.net/package/fastfetch/

**Comment by CarterLi** [comment]
I don't know much about debian and personally I don't use debian. I hope someone has experience of packaging for debian take this.

**Comment by hiagofranco** [comment]
Hi all, I would like to proceed with this task if possible. Is that ok?

I personally use Debian and I personally know some Debian developers and Debian maintainers that are used to contributing to Debian, I've already talked to one of them and they will help me. I will send an email to https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1053995.

Thank you for this nice project.

**Comment by CarterLi** [comment]
Thank you @hiagofranco! Really appreciated!

**Comment by mok0** [comment]
Also look at Maytham's comment at Bug #[1067925](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1067925). He suggests using salsa for maintaining the package. I succesfully built `fastfetch` using Carter's `debian/` but Debian projects prefer there is no `debian/` directory in the upstream tarball. If you maintain the package in git,  `debian/` only lives in the `debian/latest` branch, and the upstream source lives in the `upstream/latest` branch, look [here](https://wiki.debian.org/PackagingWithGit).

**Comment by Maytha8** [comment]
Hello @hiagofranco

> Hi all, I would like to proceed with this task if possible. Is that ok?
I've marked you as the owner of the ITP bug now.

Would you be willing to add me as a co-maintainer of this package? I'm also interested in getting this piece of software into Debian :smile:. 

I have [a bit of experience](https://qa.debian.org/developer.php?login=maytha8thedev@gmail.com) in maintaining Debian packages, and have recently become a DM, so I can help out with packaging and uploading. I can also help with organising the creation of a repo under the debian/ group on Salsa.

Have you started work and/or created a Salsa repo yet?

Thanks,
Maytham (maytham on IRC)

**Comment by hiagofranco** [comment]
Hi all,

@mok0 thanks for sharing the links.

> If you maintain the package in git, debian/ only lives in the debian/latest branch, and the upstream source lives in the upstream/latest branch, look [here](https://wiki.debian.org/PackagingWithGit).

This would be the approach, yes. I believe I will have to ignore the /debian folder from fastfetch for now and rewrite it for the Debian package, but I will post any updates here. We can probably get rid of this folder in the folder and keep it only in the Debian repo.

@Maytha8 

> I've marked you as the owner of the ITP bug now.

Thank you! I will start to work on it today.

> Would you be willing to add me as a co-maintainer of this package? I'm also interested in getting this piece of software into Debian 😄.
> I have [a bit of experience](https://qa.debian.org/developer.php?login=maytha8thedev@gmail.com) in maintaining Debian packages, and have recently become a DM, so I can help out with packaging and uploading. I can also help with organising the creation of a repo under the debian/ group on Salsa.

Of course, let's do that together, it will be nice to have someone with experience to help ;-)

> Have you started work and/or created a Salsa repo yet?

Not yet, I plan to start today, I will try to create it here and I will ask for help if something goes wrong...

How can I contact you? Can I use the email you've shared [here](https://qa.debian.org/developer.php?login=maytha8thedev@gmail.com)?

Best Regards,

Hiago.

**Comment by Maytha8** [comment]
@hiagofranco

> Not yet, I plan to start today, I will try to create it here and I will ask for help if something goes wrong...

Exciting!

> How can I contact you? Can I use the email you've shared [here](https://qa.debian.org/developer.php?login=maytha8thedev@gmail.com)?

Yep, or on IRC or GitHub.

**Comment by hiagofranco** [comment]
@Maytha8 all right, thank you =) 

I've started creating the Debian folder files and I noticed something  I believe we will have to remove. 

There is a LICENSE file for intel inside the `src/3rdparty/igcl/` folder that is not in agreement with Debian. By removing this folder, I will have to patch the source code, and `src/detection/gpu/gpu_intel.c` will not work, so any user using this GPU will not get this info. Any comments on that?

I don't know exactly how the MIT + proprietary LICENSE works as we have two licenses here, but I believe this will not be possible for the Debian package to be accepted on the mainline.

**Comment by CarterLi** [comment]
Hello, I am aware of this license issue. There are several folders in 3rd party folder and they all can be removed

1. ags and igcl: not used in Linux and can be removed directly
2. nvml: it is Nvidia's proprietary driver connector. We support a CMake option `-DENABLE_PROPRIETARY_GPU_DRIVER_API=OFF` to disable using it.
3. mk_wcwidch: it is implementation of wcwidch(3) provided for Windows copatibility. It's not used in Linux since Linux support wcwidch(3) natively
4. yyjson: this is core dependency of fastfetch to parse JSON files. It is licensed in MIT too. We also have CMake option `-DENABLE_SYSTEM_YYJSON=ON` to use system yyjson library if it is packaged officially (not the case for Debian).

**Comment by hiagofranco** [comment]
Hi @CarterLi, thank you for the detailed explanation :)

good, points 1..3 will be easy to fix then. About point 4, I have a debian contributor colleague and he will package the yyjson for Debian, as you mentioned it is not available yet but there is an ITP opened for it. Then we can make the fastfetch package dependent on the libyyjson one. I will come back with more updates onde we have it. Thanks!

**Comment by Maytha8** [comment]
@hiagofranco So I guess fastfetch will be [repacked](https://wiki.debian.org/Javascript/Repacking) to remove the 3rd-party/ directory from the Debian package?

**Comment by hiagofranco** [comment]
@Maytha8 exactly, I used the `gbp import-orig --filter=debian --filter=src/3rdparty fastfetch-2.11.3.tar.gz` to get the tarball and remove the debian and 3rdparty folders completely. Once I upload it to salsa repo I can share here with you.

**Comment by hiagofranco** [comment]
Hi @CarterLi, sorry for the absence, me, @Maytha8, and my friend @charles2910 created an IRC channel and we have been communicating there.

The fastfetch package is almost ready, it is under https://salsa.debian.org/debian/fastfetch, a repo under Debian that @Maytha8 crated. 

The major issue was that fastfetch uses libraries inside the code and this is  not recommended by Debian. Libraries should have their own package, and not be shipped inside the source code. 

So we had to first package yyjson, which is under https://salsa.debian.org/debian/yyjson. This one will soon get to the NEW queue of Debian. Then I had to remove all the other libraries due to the license issue that we discussed.

After that, we can send fastfetch and point the fastfetch to the libyyjson package and have this ready.

> In order to reduce the license concerns, I [have written my own nvml header from scratch](https://github.com/fastfetch-cli/fastfetch/blob/dev/src/detection/gpu/nvml.h) by reading [the official nvml document](https://docs.nvidia.com/deploy/nvml-api/index.html).
>I found [a program named nvtop](https://github.com/Syllo/nvtop) which was [licensed in GPLv3](https://github.com/Syllo/nvtop/blob/master/COPYING) [did the same](https://github.com/Syllo/nvtop/blob/master/src/extract_gpuinfo_nvidia.c#L38-L206). [Nvtop has been in Debian for long time](https://packages.debian.org/buster/source/nvtop) so I believe that it should be enough to make Debian mentors happy.
>There is another famous program called [btop uses the same strategy](https://github.com/aristocratos/btop/blob/3d46dc6969d1d9c981d84e81cf52b39a9f5f7941/src/linux/btop_collect.cpp#L110-L147), which is in [Gentoo](https://packages.gentoo.org/packages/sys-process/btop) for long time. CC @ceamac

Nice! =) This would be good to have the libraries available and working on Debian for sure. But this would also mean that we have to package these libraries instead of shipping them inside the fastfetch code, then we can make the fastfetch .deb as a dependency of the other libraries packages.

**Comment by CarterLi** [comment]
> But this would also mean that we have to package these libraries instead of shipping them inside the fastfetch code

Why? They didn't do it for nvtop but why is it necessary for fastfetch? It's just a header file with only function and structs declarations which generates absolutely no machine code.

**Comment by Maytha8** [comment]
> How is the debian packaging?

@CarterLi Going well, yyjson is currently in the NEW queue waiting to be accepted.

**Comment by charles2910** [comment]
FYI yyjson just got accepted!

**Comment by CarterLi** [comment]
Hello. I have cut a new release with https://github.com/fastfetch-cli/fastfetch/pull/981 included. I noticed that we still have v2.11.5 in the debian repo. I wonder if we can upgrade it to a newer version.

**Comment by hiagofranco** [comment]
Hi @CarterLi sure, we are in the middle of the review process of the Debian files before submitting it, I will be working on it today/this weekend. Then I can also update the version, thanks!

**Comment by Maytha8** [comment]
fastfetch is waiting in the NEW queue :smiley:

**Comment by AT-StephenDetomasi** [comment]
Great news, look forward to running the following on my various debian systems :)

apt remove -y neofetch && apt install -y fastfetch

**Comment by mok0** [comment]
> Great news, look forward to running the following on my various debian systems :)
> apt remove -y neofetch && apt install -y fastfetch


and in .bashrc

    alias neofetch='fastfetch --config neofetch'

:)

**Comment by Guiorgy** [comment]
> Great news, look forward to running the following on my various debian systems :)
> 
> apt remove -y neofetch && apt install -y fastfetch

You can run:

```bash
release=$(curl -s 'https://api.github.com/repos/fastfetch-cli/fastfetch/releases/latest' | grep -m 1 '"tag_name"' | cut -d : -d '"' -f 4)
wget "https://github.com/fastfetch-cli/fastfetch/releases/download/${release}-linux-amd64.deb"
sudo apt install ./${release}-linux-amd64.deb
```

To install latest version (replace amd64 if you have a different arch).

**Comment by Maytha8** [comment]
> ...if you can't wait for that.

No need to wait! fastfetch has now been accepted into Debian unstable!
https://tracker.debian.org/pkg/fastfetch

@CarterLi I reckon this issue can be closed now

**Comment by CarterLi** [comment]
Cool! Big thanks!

**Comment by nolddor** [comment]
> > ...if you can't wait for that.
> 
> No need to wait! fastfetch has now been accepted into Debian unstable! https://tracker.debian.org/pkg/fastfetch
> 
> @CarterLi I reckon this issue can be closed now

Accepted into Debian testing as well ! Congrats

**Comment by mok0** [comment]
This is indeed great, congrats to all involved! I now see that fastfetch is already in Testing.

However, please allow my to comment on the package Description. This is how it looks like a.t.m.:

> Description: neofetch-like tool for fetching system information
>  Fastfetch is a neofetch-like tool for fetching system information and
>  displaying them in a pretty way. It is written mainly in C, with
>  performance and customizability in mind.

I don't think it's good practice to refer to another program in the description section, in this case neofetch. The description assumes that the user knows what neofetch is, and given that it is abandoned by the author and likely  going away in the future, users will soon not know what neofetch is.

I think it would be better to refer to neofetch in the man page, I am not sure to what extent fastfetch was inspired by neofetch, but perhaps it would be proper to credit the author of neofetch in the man page, if this is the case. Neofetch has been a great program and used by many, and now we say goodbye to it and hello to fastfetch :-)

I have authored an alternative Description: text that you are free to use in any way you please, as you can see, I have avoided mentioning neofetch:

> Description: Pretty display of detailed system information
>  Fastfetch is a highly configurable, cross-platform tool that fetches
>  system information and displays it on the terminal in a pretty way.
>  Fastfetch comes with over 60 modules and over 400 logos that can
>  optionally be displayed when configured by the user. Fastfetch is
>  mainly written in C, with performance and customizability in mind.

**Comment by CarterLi** [comment]
Hello @Maytha8. The debian package seems stucked at 2.15. Is there any plan to upgrade it? What can I help?

**Comment by Guiorgy** [comment]
@CarterLi It was updated yesterday to [2.21.3+dfsg-1](https://packages.debian.org/source/testing/fastfetch) 🎆

**Comment by thesolomon-tech** [comment]
Hi, it seems that Fastfetch is not available on the latest versions of Ubuntu. Might be good to reopen this issue.

**Comment by Guiorgy** [comment]
> Hi, it seems that Fastfetch is not available on the latest versions of Ubuntu. It might be good to reopen this issue.

What do you mean? [Oracular (24.10) has 2.15.0](https://packages.ubuntu.com/oracular/fastfetch), and [Plucky (25.04) has 2.38.0](https://packages.ubuntu.com/plucky/fastfetch).

**Comment by Guiorgy** [comment]
> [...] I think it'll be added in the next LTS release, though I'm unsure as I'm not familiar with Ubuntu development.

Should be, unless it is dropped.

**Comment by thesolomon-tech** [comment]
> > Hi, it seems that Fastfetch is not available on the latest versions of Ubuntu. It might be good to reopen this issue.
> 
> It's in 24.10, but not in 24.04 (LTS). I think it'll be added in the next LTS release, though I'm unsure as I'm not familiar with Ubuntu development.

That explains it. I am using the latest LTS version and it is not present.

**Comment by AT-StephenDetomasi** [comment]
It's also not available in Proxmox 8.4. I'm surprised it's not available yet.
